openpathsampling/experimental/storage/storage.py

```diff
@@ -0,0 +1,531 @@
+        return out_str
+
+    def __getattr__(self, attr):
+        # override getattr to create iterators over the tables (stores)
+        if attr in self._storage_tables:
+            return self._storage_tables[attr]
+        elif attr in self._pseudo_tables:
+            return self._pseudo_tables[attr]
+        else:
+            raise AttributeError("'{}' object has no attribute '{}'"\
+                                 .format(self.__class__.__name__, attr))
+
+
+class MixedCache(abc.MutableMapping):
+    """Combine a frozen cache and a mutable cache"""
+    # TODO: benchmark with single dict instead; might be just as fast!
+    def __init__(self, fixed_cache=None):
+        self.fixed_cache = tools.none_to_default(fixed_cache, default={})
+        self.cache = {}
+
+    def clear(self):
+        self.cache = {}
+
+    def delete_items(self, list_of_items, error_if_missing=False):
+        for item in list_of_items:
+            if item in self:
+                del self[item]
+            elif error_if_missing:
+                raise KeyError()  # TODO: message and check error type
+
+    def reproxy(self, schema):
+        # TODO: idea: turn loaded objects back into None for any proxy
+        # objects in cache. This frees those things up for garbage
+        # collection.
+        pass
+
+    def __getitem__(self, key):
+        if key in self.fixed_cache:
+            value = self.fixed_cache[key]
+        else:
+            value = self.cache[key]
+        return value
+
+    def __setitem__(self, key, value):
+        self.cache[key] = value
+
+    def __delitem__(self, key):
+        try:
+            del self.cache[key]
+        except KeyError as err:
+            if key in self.fixed_cache:
+                raise TypeError("Can't delete from fixed cache")
+            else:
+                raise err
+
+    def __len__(self):
+        return len(self.fixed_cache) + len(self.cache)
+
+    def __iter__(self):
+        return itertools.chain(self.fixed_cache, self.cache)
+
+class StorageTable(abc.Sequence):
+    # NOTE: currently you still need to be able to hold the whole table in
+    # memory ... at least, with the SQL backend.
+    def __init__(self, storage, table):
+        self.storage = storage
+        self.table = table
+        self.clear_cache_block_freq = 100
+        self.iter_block_size = 100  # TODO: base it on the size of an object
+
+    def __iter__(self):
+        # TODO: ensure that this gives us things in idx order
+        backend_iter = self.storage.backend.table_iterator(self.table)
+        enum_iter = enumerate(tools.grouper(backend_iter,
+                                            self.iter_block_size))
+        for block_num, block in enum_iter:
+            row_uuids = [row.uuid for row in block]
+            loaded = self.storage.load(row_uuids)
+            if block_num % self.clear_cache_block_freq == 0:
+                self.storage.cache.clear()
```

Comment:
This works for the new analysis, but the old one errors out ([details on how this was figured out](https://github.com/dwhswenson/ops-storage-notebooks/issues/1)) :

If you want to support the old analysis for a bit longer, this can be changed to:
```suggestion
            row_uuids = [row.uuid for row in block]
            loaded = self.storage.load(row_uuids)
            if block_num % self.clear_cache_block_freq == 0 and block_num != 0:
                self.storage.cache.clear()
```

or

```suggestion
            row_uuids = [row.uuid for row in block]
            if block_num % self.clear_cache_block_freq == 0:
                self.storage.cache.clear()
            loaded = self.storage.load(row_uuids)
```

Reply:
Thanks!